### DISCUSSION TITLE
starlette.TestClient with Jinja2Templates and middleware

### DISCUSSION THREAD

**Discussion by bloomingmath** [fastapi/fastapi#7865] [Questions] [answered]
### Describe the bug

The problem arise when using starlette TestClient to get a template response that passes through a middleware.

### To Reproduce

I made a repo with the minimal fastapi app to raise the error:

https://github.com/bloomingmath/fastapi-middleware-issue

To reproduce, clone the repo and run `pytest main.py`

### Expected behavior

I expect the test to run flawlessly, but it raise AssertionError as shown in the readme.md of the same repo.

### Environment

- OS: Ubuntu 18.04
- Python 3.7.5
- attrs==19.3.0
- certifi==2019.11.28
- chardet==3.0.4
- fastapi==0.45.0
- idna==2.8
- importlib-metadata==1.3.0
- Jinja2==2.10.3
- MarkupSafe==1.1.1
- more-itertools==8.0.2
- packaging==19.2
- pluggy==0.13.1
- py==1.8.0
- pydantic==1.2
- pyparsing==2.4.5
- pytest==5.3.2
- requests==2.22.0
- six==1.13.0
- starlette==0.12.9
- urllib3==1.25.7
- wcwidth==0.1.7
- zipp==0.6.0

**Comment by podhmo** [comment]
This is related issue, in starlette, maybe...

https://github.com/encode/starlette/issues/472

Hmm.. this is strange, maybe. Somehow, message["type"] is not "http.response.start", in starlette's middleware.
I print debugged, this value is "http.response.template".

actual value is like below.

```
{'type': 'http.response.template', 'template': <Template 'page.html'>, 'context': {'request': <starlette.requests.Request object at 0x7f9f2ced5850>}}
```

So, assertion error occured in this line

https://github.com/encode/starlette/blob/00e830080499c47a3e72f1b4adb2678758e388ee/starlette/middleware/base.py#L47

## but async-asgi-testclient is ok

Writing similar tests, using [async-asgi-testclient](https://github.com/vinissimus/async-asgi-testclient) , test is passed.

```python
import pytest


@pytest.mark.asyncio
async def test_middleware2():
    from async_asgi_testclient import TestClient

    async with TestClient(app) as client:
        response = await client.get("/")
        assert "X-Process-Time" in response.headers
```

then, message value is below

```
{'type': 'http.response.start', 'status': 200, 'headers': [(b'content-length', b'131'), (b'content-type', b'text/html; charset=utf-8')]}
```

**Comment by bloomingmath** [comment]
You are right @podhmo, the problem is with starlette and not with fastapi. Also, using async-asgi-testclient permits testing the application until starlette's test client is fixed.
Thank you for your help.

**Comment by podhmo** [comment, marked as answer]
OK. I understood completely. This is the same issue as https://github.com/encode/starlette/issues/472

1. Starlette's template response class treats specially when scope has `{ "extensions": { "http.response.template": ...}}`, then returns "http.response.template" message. ( [here](https://github.com/encode/starlette/blob/65b280fcb590c69d0ed44b8452522b5277a19e7a/starlette/templating.py#L33-L40) )
2. when Starlette's testclient sendindg http request, always attaching data including `{ "extensions": { "http.response.template": ...}}` ( [here](https://github.com/encode/starlette/blob/65b280fcb590c69d0ed44b8452522b5277a19e7a/starlette/testclient.py#L154) )
3. And, BaseMiddleware asserting "http.response.start" message. ( [here](https://github.com/encode/starlette/blob/65b280fcb590c69d0ed44b8452522b5277a19e7a/starlette/middleware/base.py#L47) )

So, test is failed.

**Comment by tiangolo** [comment]
Thanks for the help here @podhmo ! :clap: :bow: 

Thanks for reporting back and closing the issue @bloomingmath  :+1:

**Comment by Jdsleppy** [comment]
Hi all, it would be wonderful to be able to read and assert on `response.context` or `response.template` in unit tests.  I can tell that this work and Starlette's `TestClient` are _trying_ to make that possible, but I end up not seeing these attributes on my responses.

A non-working example test would be:

```python
from fastapi import FastAPI, status
from fastapi.testclient import TestClient

class TestSocList(object):
    async def test_empty(self, app: FastAPI, client: TestClient):
        response = client.get(app.url_path_for("my_route"))
        assert response.status_code == status.HTTP_200_OK
        assert response.context
```

Am I missing something?  I see the `http.response.debug` message get sent but I think it's not propagating up through all of the middleware to the `TestClient`.

**Reply by Jdsleppy** [reply]
It looks like Starlette base middlewares both:

1. throw away the debug message https://github.com/encode/starlette/blob/5f9da0b6564e1b7def4d002f222766db5090735c/starlette/middleware/base.py#L161, and
2. throw away the `TemplateResponse` and return a new response object https://github.com/encode/starlette/blob/5f9da0b6564e1b7def4d002f222766db5090735c/starlette/middleware/base.py#L183

So I'm not sure there's any easy way to get that debug info out.

But wait, it passes along the `info` which is a `dict` containing the context and template:

```python
response = _StreamingResponse(
    status_code=message["status"], content=body_stream(), info=info
)
```

And it tries to send it along https://github.com/encode/starlette/blob/5f9da0b6564e1b7def4d002f222766db5090735c/starlette/middleware/base.py#L216

But I'm using the `GZipMiddleware` which drops the message entirely if it's of type `http.response.debug`! https://github.com/encode/starlette/blob/5f9da0b6564e1b7def4d002f222766db5090735c/starlette/middleware/gzip.py#L48

Comment out the `GZipMiddleware` and it works like a charm!
